odata `groupBy` causes `subquery uses ungrouped column "__select__.author_id" from outer query`

### Description of erroneous behaviour 

CDS server crashes with error `TypeError: transformedQuery.SELECT.from.ref is not iterable` when trying to `groupBy` on a child association.

Stacktrace:
```
[cds] - TypeError: transformedQuery.SELECT.from.ref is not iterable
    at expandColumn (node_modules\@cap-js\db-service\lib\cqn4sql.js:790:41)
    at handleExpand (node_modules\@cap-js\db-service\lib\cqn4sql.js:420:40)
    at node_modules\@cap-js\db-service\lib\cqn4sql.js:354:32
    at getTransformedColumns (node_modules\@cap-js\db-service\lib\cqn4sql.js:382:21)
    at transformSelectQuery (node_modules\@cap-js\db-service\lib\cqn4sql.js:156:41)
    at cqn4sql (node_modules\@cap-js\db-service\lib\cqn4sql.js:73:26)
    at PostgresService.cqn4sql (node_modules\@cap-js\db-service\lib\SQLService.js:375:12)
    at PostgresService.cqn2sql (node_modules\@cap-js\db-service\lib\SQLService.js:351:18)
    at PostgresService.onSELECT (node_modules\@cap-js\db-service\lib\SQLService.js:127:39)
    at PostgresService.onSELECT (node_modules\@cap-js\postgres\lib\PostgresService.js:314:18) {
  id: '1043785',
  level: 'ERROR',
  timestamp: 1719308262527
}
```

### Detailed steps to reproduce

Try to run an odata query similar to the following:
```
$apply: filter(foo eq 'bar')/groupby((user/name),aggregate(amount with sum as totalAmount))
```
If you change `groupby((user/name)` with `groupby((user_ID)` it works as expected.

### Details about your project

| Package | Version |
|:------------------|---------------------------------------|
| OData version     | v4                               |
| @cap-js/cds-typer      | 0.22.0                                                     |
| @cap-js/cds-types      | 0.2.0                                                      |
| @cap-js/postgres       | 1.8.0                                                      |
| @sap/cds               | 7.9.2                                                      |
| @sap/cds-compiler      | 4.9.4                                                      |
| @sap/cds-dk            | 7.9.3                                                      |
| @sap/cds-dk (global)   | 7.9.2                                                      |
| @sap/cds-fiori         | 1.2.4                                                      |
| @sap/cds-foss          | 5.0.1                                                      |
| @sap/cds-mtxs          | 1.18.1                                                     |
| @sap/eslint-plugin-cds | 3.0.4                                                      |
| Node.js                | v20.13.1                                                   |

compiling your model with `cds compile -2 sql` does not work:

```
❯ cds compile -2 sql e.cds
[ERROR] e.cds:35:13: “V”: “isRecurring” is not foreign key of managed association “product” in path “orderProduct.product.isRecurring” (in entity:“V”/query:1)
[ERROR] e.cds:35:13: “V”: “product” is not foreign key of managed association “orderProduct” in path “orderProduct.product.isRecurring” (in entity:“V”/query:1)
```

---

The calculated element is not valid. That is because it navigates to a non-foreign-key in path `orderProduct.product.isRecurring` where the association `product` is not fk of `orderProduct` and `isRecurring` is not a fk element of `product`.

This is a known limitation. The filter path along `contracts[…].orderProduct` induces a join by itself, with `[…]` being part of the `join` node.

this would be the resulting view if each element along the path in question would be a foreign key element:
```sql
CREATE VIEW V AS
SELECT Customers_0.ID,
  sum(orderProduct_2.price) AS totalValue
FROM (
    (
      Customers AS Customers_0
      LEFT JOIN Contracts AS contracts_1 ON ((contracts_1.customer_ID = Customers_0.ID))
      AND (
        contracts_1.status = 'Active'
        AND contracts_1.orderProduct_product_isRecurring = TRUE
      )
    )
    LEFT JOIN OrderProducts AS orderProduct_2 ON (contracts_1.orderProduct_ID = orderProduct_2.ID)
    AND (
      contracts_1.orderProduct_product_ID = orderProduct_2.product_ID
    )
    AND (
      contracts_1.orderProduct_product_isRecurring = orderProduct_2.product_isRecurring
    )
  )
```

→ The filter path is part of the join. If the filter condition itself induces joins (due to non-foreign-key relations), things get much more complex.

--- a/db-service/lib/cqn2sql.js
+++ b/db-service/lib/cqn2sql.js
@@ -222,6 +222,7 @@ class CQN2SQLRenderer {
     if (distinct) sql += ` DISTINCT`
     if (!_empty(columns)) sql += ` ${columns}`
     if (!_empty(from)) sql += ` FROM ${this.from(from)}`
+    else sql += this.from_dummy()
     if (!_empty(where)) sql += ` WHERE ${this.where(where)}`
     if (!_empty(groupBy)) sql += ` GROUP BY ${this.groupBy(groupBy)}`
     if (!_empty(having)) sql += ` HAVING ${this.having(having)}`
@@ -344,6 +345,14 @@ class CQN2SQLRenderer {
     if (from.SELECT) return _aliased(`(${this.SELECT(from)})`)
     if (from.join)
       return `${this.from(from.args[0])} ${from.join} JOIN ${this.from(from.args[1])} ON ${this.where(from.on)}`
+  }
+
+  /**
+   * Renders a FROM clause for when the query does not have a target
+   * @returns {string} SQL
+   */
+  from_dummy() {
+    return ''
   }
 
   /**

--- a/db-service/lib/cqn4sql.js
+++ b/db-service/lib/cqn4sql.js
@@ -798,6 +798,7 @@ function cqn4sql(originalQuery, model) {
       (column.$refLinks[0].definition.kind === 'entity'
         ? column.ref.slice(1).map(idOnly).join('_') // omit explicit table alias from name of column
         : column.ref.map(idOnly).join('_'))
+
     // if there is a group by on the main query, all
     // columns of the expand must be in the groupBy
     if (transformedQuery.SELECT.groupBy) {
@@ -861,8 +862,22 @@ function cqn4sql(originalQuery, model) {
       return subq
     }
 
+    /**
+     * Generates a special subquery for the `expand` of the `column`.
+     * All columns in the `expand` must be part of the GROUP BY clause of the main query.
+     * If this is the case, the subqueries columns match the corresponding references of the group by.
+     * Nested expands are also supported.
+     *
+     * @param {Object} column - To expand.
+     * @param {Array} baseRef - The base reference for the expanded column.
+     * @param {string} subqueryAlias - The alias of the `expand` subquery column.
+     * @returns {Object} - The subquery object.
+     * @throws {Error} - If one of the `ref`s in the `column.expand` is not part of the GROUP BY clause.
+     */
     function _subqueryForGroupBy(column, baseRef, subqueryAlias) {
-      const groupByLookup = new Map(transformedQuery.SELECT.groupBy.map(c => [c.ref && c.ref.map(idOnly).join('.'), c]))
+      const groupByLookup = new Map(
+        transformedQuery.SELECT.groupBy.map(c => [c.ref && c.ref.map(refWithConditions).join('.'), c]),
+      )
 
       // to be attached to dummy query
       const elements = {}
@@ -870,14 +885,16 @@ function cqn4sql(originalQuery, model) {
         const fullRef = [...baseRef, ...expand.ref]
 
         if (expand.expand) {
-          subqueryAlias = expand.as || expand.ref.map(idOnly).join('_')
-          const nested = _subqueryForGroupBy(expand, fullRef, subqueryAlias)
+          const nested = _subqueryForGroupBy(expand, fullRef, expand.as || expand.ref.map(idOnly).join('_'))
+          elements[expand.as || expand.ref.map(idOnly).join('_')] = nested
           return nested
         }
 
-        const groupByRef = groupByLookup.get(fullRef.join('.'))
+        const groupByRef = groupByLookup.get(fullRef.map(refWithConditions).join('.'))
         if (!groupByRef) {
-          throw new Error(`"${fullRef.join('.')}": The expanded column must be part of the group by clause.`)
+          throw new Error(
+            `The expanded column "${fullRef.map(refWithConditions).join('.')}" must be part of the group by clause`,
+          )
         }
 
         const columnCopy = Object.create(groupByRef)
@@ -889,21 +906,28 @@ function cqn4sql(originalQuery, model) {
           const name = calculateElementName(columnCopy)
           columnCopy.ref = [tableAlias, name]
         }
+        elements[expand.as || expand.ref.map(idOnly).join('_')] = columnCopy.$refLinks.at(-1).definition
         return columnCopy
       })
 
-      const SELECT = {}
-      Object.defineProperties(SELECT, {
-        from: { value: null, enumerable: true },
-        columns: { value: expandedColumns, enumerable: true },
-        expand: { value: true },
-        one: { value: column.$refLinks.at(-1).definition.is2one },
-      })
-
-      return {
-        SELECT,
-        as: subqueryAlias,
+      const SELECT = {
+        from: null,
+        columns: expandedColumns,
       }
+      return Object.defineProperties(
+        {},
+        {
+          SELECT: {
+            value: Object.defineProperties(SELECT, {
+              expand: { value: true }, // non-enumerable
+              one: { value: column.$refLinks.at(-1).definition.is2one }, // non-enumerable
+            }),
+            enumerable: true,
+          },
+          as: { value: subqueryAlias, enumerable: true },
+          elements: { value: elements }, // non-enumerable
+        },
+      )
     }
   }
 
@@ -2305,4 +2329,12 @@ function setElementOnColumns(col, element) {
 
 const getName = col => col.as || col.ref?.at(-1)
 const idOnly = ref => ref.id || ref
+const refWithConditions = step => {
+  let appendix
+  const { args, where } = step
+  if (where && args) appendix = JSON.stringify(where) + JSON.stringify(args)
+  else if (where) appendix = JSON.stringify(where)
+  else if (args) appendix = JSON.stringify(args)
+  return appendix ? step.id + appendix : step
+}
 const is_regexp = x => x?.constructor?.name === 'RegExp' // NOTE: x instanceof RegExp doesn't work in repl

--- a/db-service/lib/cqn4sql.js
+++ b/db-service/lib/cqn4sql.js
@@ -919,12 +919,12 @@ function cqn4sql(originalQuery, model) {
         {
           SELECT: {
             value: Object.defineProperties(SELECT, {
-              expand: { value: true }, // non-enumerable
-              one: { value: column.$refLinks.at(-1).definition.is2one }, // non-enumerable
+              expand: { value: true, writable: true }, // non-enumerable
+              one: { value: column.$refLinks.at(-1).definition.is2one, writable: true }, // non-enumerable
             }),
             enumerable: true,
           },
-          as: { value: subqueryAlias, enumerable: true },
+          as: { value: subqueryAlias, enumerable: true, writable: true },
           elements: { value: elements }, // non-enumerable
         },
       )

--- a/db-service/lib/cqn4sql.js
+++ b/db-service/lib/cqn4sql.js
@@ -798,44 +798,20 @@ function cqn4sql(originalQuery, model) {
       (column.$refLinks[0].definition.kind === 'entity'
         ? column.ref.slice(1).map(idOnly).join('_') // omit explicit table alias from name of column
         : column.ref.map(idOnly).join('_'))
-
     // if there is a group by on the main query, all
     // columns of the expand must be in the groupBy
     if (transformedQuery.SELECT.groupBy) {
-      const groupByLookup = new Set(transformedQuery.SELECT.groupBy.map(c => c.ref && c.ref.map(idOnly).join('.')))
       const baseRef =
         column.$refLinks[0].definition.SELECT || column.$refLinks[0].definition.kind === 'entity'
           ? column.ref.slice(1)
           : column.ref
-      // to be attached to dummy query
-      const elements = {}
-      const expandedColumns = column.expand.map(col => {
-        const fullRef = [...baseRef, ...col.ref]
 
-        if (!groupByLookup.has(fullRef.join('.'))) {
-          throw new Error(`"${fullRef.join('.')}": The expanded column must be part of the group by clause.`)
-        }
-
-        const columnCopy = { ref: fullRef, as: col.as || undefined }
-        return columnCopy
-      })
-
-      const SELECT = {}
-      Object.defineProperties(SELECT, {
-        from: { value: { ref: ['DUMMY'] }, enumerable: true },
-        columns: { value: expandedColumns, enumerable: true },
-        expand: { value: true },
-        one: { value: column.$refLinks.at(-1).definition.is2one },
-      })
-      return {
-        SELECT,
-        as: columnAlias,
-      }
+      return _subqueryForGroupBy(column, baseRef, columnAlias)
     }
+
     // we need to respect the aliases of the outer query, so the columnAlias might not be suitable
     // as table alias for the correlated subquery
     const uniqueSubqueryAlias = getNextAvailableTableAlias(columnAlias, originalQuery.outerQueries)
-
     // `SELECT from Authors {  books.genre as genreOfBooks { name } } becomes `SELECT from Books:genre as genreOfBooks`
     const from = { ref: subqueryFromRef, as: uniqueSubqueryAlias }
     const subqueryBase = Object.fromEntries(
@@ -883,6 +859,51 @@ function cqn4sql(originalQuery, model) {
         return x
       }
       return subq
+    }
+
+    function _subqueryForGroupBy(column, baseRef, subqueryAlias) {
+      const groupByLookup = new Map(transformedQuery.SELECT.groupBy.map(c => [c.ref && c.ref.map(idOnly).join('.'), c]))
+
+      // to be attached to dummy query
+      const elements = {}
+      const expandedColumns = column.expand.map(expand => {
+        const fullRef = [...baseRef, ...expand.ref]
+
+        if (expand.expand) {
+          subqueryAlias = expand.as || expand.ref.map(idOnly).join('_')
+          const nested = _subqueryForGroupBy(expand, fullRef, subqueryAlias)
+          return nested
+        }
+
+        const groupByRef = groupByLookup.get(fullRef.join('.'))
+        if (!groupByRef) {
+          throw new Error(`"${fullRef.join('.')}": The expanded column must be part of the group by clause.`)
+        }
+
+        const columnCopy = Object.create(groupByRef)
+        if (expand.as) {
+          columnCopy.as = expand.as
+        }
+        if (columnCopy.isJoinRelevant) {
+          const tableAlias = getQuerySourceName(columnCopy)
+          const name = calculateElementName(columnCopy)
+          columnCopy.ref = [tableAlias, name]
+        }
+        return columnCopy
+      })
+
+      const SELECT = {}
+      Object.defineProperties(SELECT, {
+        from: { value: null, enumerable: true },
+        columns: { value: expandedColumns, enumerable: true },
+        expand: { value: true },
+        one: { value: column.$refLinks.at(-1).definition.is2one },
+      })
+
+      return {
+        SELECT,
+        as: subqueryAlias,
+      }
     }
   }
 

--- a/db-service/lib/cqn4sql.js
+++ b/db-service/lib/cqn4sql.js
@@ -799,6 +799,39 @@ function cqn4sql(originalQuery, model) {
         ? column.ref.slice(1).map(idOnly).join('_') // omit explicit table alias from name of column
         : column.ref.map(idOnly).join('_'))
 
+    // if there is a group by on the main query, all
+    // columns of the expand must be in the groupBy
+    if (transformedQuery.SELECT.groupBy) {
+      const groupByLookup = new Set(transformedQuery.SELECT.groupBy.map(c => c.ref && c.ref.map(idOnly).join('.')))
+      const baseRef =
+        column.$refLinks[0].definition.SELECT || column.$refLinks[0].definition.kind === 'entity'
+          ? column.ref.slice(1)
+          : column.ref
+      // to be attached to dummy query
+      const elements = {}
+      const expandedColumns = column.expand.map(col => {
+        const fullRef = [...baseRef, ...col.ref]
+
+        if (!groupByLookup.has(fullRef.join('.'))) {
+          throw new Error(`"${fullRef.join('.')}": The expanded column must be part of the group by clause.`)
+        }
+
+        const columnCopy = { ref: fullRef, as: col.as || undefined }
+        return columnCopy
+      })
+
+      const SELECT = {}
+      Object.defineProperties(SELECT, {
+        from: { value: { ref: ['DUMMY'] }, enumerable: true },
+        columns: { value: expandedColumns, enumerable: true },
+        expand: { value: true },
+        one: { value: column.$refLinks.at(-1).definition.is2one },
+      })
+      return {
+        SELECT,
+        as: columnAlias,
+      }
+    }
     // we need to respect the aliases of the outer query, so the columnAlias might not be suitable
     // as table alias for the correlated subquery
     const uniqueSubqueryAlias = getNextAvailableTableAlias(columnAlias, originalQuery.outerQueries)
@@ -815,7 +848,7 @@ function cqn4sql(originalQuery, model) {
         from,
         columns: JSON.parse(JSON.stringify(column.expand)),
         expand: true,
-        one: column.$refLinks[column.$refLinks.length - 1].definition.is2one,
+        one: column.$refLinks.at(-1).definition.is2one,
       },
     }
     const expanded = transformSubquery(subquery)

--- a/db-service/test/cqn4sql/assocs2joins.test.js
+++ b/db-service/test/cqn4sql/assocs2joins.test.js
@@ -728,18 +728,15 @@ describe('Unfolding Association Path Expressions to Joins', () => {
   it('properly rewrite association chains if intermediate assoc is not fk', () => {
     // this issue came up for ref: [genre.parent.ID] because "ID" is fk of "parent"
     // but "parent" is not fk of "genre"
-    const q = CQL`SELECT from (select genre from bookshop.Books) as book {
-      genre { name }
+    const q = CQL`SELECT from (select genre, ID from bookshop.Books) as book {
+      ID
     } group by genre.parent.ID, genre.parent.name`
     const qx = CQL`
-    SELECT from (select Books.genre_ID from bookshop.Books as Books) as book
+    SELECT from (select Books.genre_ID, Books.ID from bookshop.Books as Books) as book
                                 left join bookshop.Genres as genre on genre.ID = book.genre_ID
                                 left join bookshop.Genres as parent on parent.ID = genre.parent_ID
     {
-      (SELECT
-          genre2.name
-        from bookshop.Genres as genre2
-        where book.genre_ID = genre2.ID) as genre
+      book.ID
     } group by parent.ID, parent.name`
     const res = cqn4sql(q, model)
     expect(JSON.parse(JSON.stringify(res))).to.deep.eql(qx)

--- a/db-service/test/cqn4sql/expand.test.js
+++ b/db-service/test/cqn4sql/expand.test.js
@@ -1046,6 +1046,7 @@ describe('Expands with aggregations are special', () => {
       Books.ID,
       (SELECT from DUMMY { author.name }) as author
     } group by author.name`
+    qx.SELECT.columns[1].SELECT.from = null
     const res = cqn4sql(q, model)
     expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
   })
@@ -1064,23 +1065,53 @@ describe('Expands with aggregations are special', () => {
       (SELECT from DUMMY { author.name }) as author,
       (SELECT from DUMMY { genre.name }) as genre
     } group by author.name, genre.name`
+    qx.SELECT.columns[1].SELECT.from = null
+    qx.SELECT.columns[2].SELECT.from = null
     const res = cqn4sql(q, model)
     expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
   })
-  it.skip('with nested expands', () => {
+  it('with nested expands', () => {
     const q = CQL`SELECT from bookshop.Genres {
       ID,
       Genres.parent { parent { name } },
     } group by parent.parent.name`
 
-    const qx = CQL`SELECT from bookshop.Books as Books
-    left join bookshop.Authors as author on author.ID = Books.author_ID
-    left join bookshop.Genres as genre on genre.ID = Books.genre_ID
+    const qx = CQL`SELECT from bookshop.Genres as Genres
+    left join bookshop.Genres as parent on parent.ID = Genres.parent_ID
+    left join bookshop.Genres as parent2 on parent2.ID = parent.parent_ID
     {
-      Books.ID,
-      (SELECT from DUMMY { author.name }) as author,
-      (SELECT from DUMMY { genre.name }) as genre
-    } group by author.name, genre.name`
+      Genres.ID,
+      (
+        SELECT from DUMMY {
+          (SELECT from DUMMY { parent2.name }) as parent
+        }
+      ) as parent,
+    } group by parent2.name`
+    qx.SELECT.columns[1].SELECT.from = null
+    qx.SELECT.columns[1].SELECT.columns[0].SELECT.from = null
+    const res = cqn4sql(q, model)
+    expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
+  })
+  it('with nested expands and non-nested sibling', () => {
+    const q = CQL`SELECT from bookshop.Genres {
+      ID,
+      Genres.parent { parent { name }, name },
+    } group by parent.parent.name, parent.name`
+
+    const qx = CQL`SELECT from bookshop.Genres as Genres
+    left join bookshop.Genres as parent on parent.ID = Genres.parent_ID
+    left join bookshop.Genres as parent2 on parent2.ID = parent.parent_ID
+    {
+      Genres.ID,
+      (
+        SELECT from DUMMY {
+          (SELECT from DUMMY { parent2.name }) as parent,
+          parent.name
+        }
+      ) as parent,
+    } group by parent2.name, parent.name`
+    qx.SELECT.columns[1].SELECT.from = null
+    qx.SELECT.columns[1].SELECT.columns[0].SELECT.from = null
     const res = cqn4sql(q, model)
     expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
   })

--- a/db-service/test/cqn4sql/expand.test.js
+++ b/db-service/test/cqn4sql/expand.test.js
@@ -194,7 +194,6 @@ describe('Unfold expands on associations to special subselects', () => {
     expect(() => cqn4sql(CQL`SELECT from bookshop.Books { author[books.title = 'foo'] { name } }`, model)).to.throw(
       /Unexpected unmanaged association “books” in filter expression of “author”/,
     )
-
   })
   it('rejects non-fk access in infix filter of expand path', () => {
     expect(() =>
@@ -280,7 +279,6 @@ describe('Unfold expands on associations to special subselects', () => {
     }`
     expect(JSON.parse(JSON.stringify(res))).to.deep.equal(expected)
   })
-
 
   it('add where exists <assoc> shortcut to expand subquery where condition', () => {
     const q = {
@@ -1031,6 +1029,68 @@ describe('Unfold expands on associations to special subselects', () => {
     `)
   })
 })
+
+describe('Expands with aggregations are special', () => {
+  let model
+  beforeAll(async () => {
+    cds.model = model = await cds.load(__dirname + '/../bookshop/db/schema').then(cds.linked)
+  })
+
+  it('simple aggregation', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      ID,
+      Books.author { name }
+    } group by author.name`
+
+    const qx = CQL`SELECT from bookshop.Books as Books left join bookshop.Authors as author on author.ID = Books.author_ID {
+      Books.ID,
+      (SELECT from DUMMY { author.name }) as author
+    } group by author.name`
+    const res = cqn4sql(q, model)
+    expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
+  })
+  it('with multiple expands', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      ID,
+      Books.author { name },
+      genre { name }
+    } group by author.name, genre.name`
+
+    const qx = CQL`SELECT from bookshop.Books as Books
+    left join bookshop.Authors as author on author.ID = Books.author_ID
+    left join bookshop.Genres as genre on genre.ID = Books.genre_ID
+    {
+      Books.ID,
+      (SELECT from DUMMY { author.name }) as author,
+      (SELECT from DUMMY { genre.name }) as genre
+    } group by author.name, genre.name`
+    const res = cqn4sql(q, model)
+    expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
+  })
+  it.skip('with nested expands', () => {
+    const q = CQL`SELECT from bookshop.Genres {
+      ID,
+      Genres.parent { parent { name } },
+    } group by parent.parent.name`
+
+    const qx = CQL`SELECT from bookshop.Books as Books
+    left join bookshop.Authors as author on author.ID = Books.author_ID
+    left join bookshop.Genres as genre on genre.ID = Books.genre_ID
+    {
+      Books.ID,
+      (SELECT from DUMMY { author.name }) as author,
+      (SELECT from DUMMY { genre.name }) as genre
+    } group by author.name, genre.name`
+    const res = cqn4sql(q, model)
+    expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
+  })
+
+  it.skip('what about the filter condition of the expand?', () => {
+    CQL`SELECT from bookshop.Books {
+      author[name='King'] { name }
+    } group by author.name`
+  })
+})
 // the tests in here are a copy of the tests in `./inline.test.js`
 // and should behave exactly the same.
 // `.inline` and `.expand` on a `struct` are semantically equivalent.
@@ -1108,7 +1168,9 @@ describe('expand on structure part II', () => {
         (SELECT department.id, department.name from Department as department where Employee.department_id = department.id) as department,
         (SELECT assets.id, assets.descr from Assets as assets where Employee.id = assets.owner_id) as assets
     }`
-    expect(JSON.parse(JSON.stringify(cqn4sql(expandQuery, cds.compile.for.nodejs(JSON.parse(JSON.stringify(model))))))).to.eql(expected)
+    expect(
+      JSON.parse(JSON.stringify(cqn4sql(expandQuery, cds.compile.for.nodejs(JSON.parse(JSON.stringify(model)))))),
+    ).to.eql(expected)
   })
 
   it('multi expand with star but foreign key does not survive in structured mode', () => {

--- a/db-service/test/cqn4sql/expand.test.js
+++ b/db-service/test/cqn4sql/expand.test.js
@@ -1050,6 +1050,22 @@ describe('Expands with aggregations are special', () => {
     const res = cqn4sql(q, model)
     expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
   })
+  it('expand path with filter must be an exact match in group by', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      Books.ID,
+      author[name='King'] { name }
+    } group by author[name='King'].name`
+
+    const qx = CQL`SELECT from bookshop.Books as Books
+    left join bookshop.Authors as author on author.ID = Books.author_ID and author.name = 'King' {
+      Books.ID,
+      (SELECT from DUMMY { author.name }) as author
+    } group by author.name`
+    qx.SELECT.columns[1].SELECT.from = null
+    const res = cqn4sql(q, model)
+    expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
+  })
+
   it('with multiple expands', () => {
     const q = CQL`SELECT from bookshop.Books {
       ID,
@@ -1116,10 +1132,51 @@ describe('Expands with aggregations are special', () => {
     expect(JSON.parse(JSON.stringify(res))).to.deep.equal(qx)
   })
 
-  it.skip('what about the filter condition of the expand?', () => {
-    CQL`SELECT from bookshop.Books {
+  // negative tests
+  it('simple path not part of group by', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      ID,
+      Books.author { name, ID }
+    } group by author.name`
+
+    expect(() => cqn4sql(q, model)).to.throw(/The expanded column "author.ID" must be part of the group by clause/)
+  })
+  it('nested path not part of group by', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      ID,
+      Books.author { books {title}, ID }
+    } group by author.ID`
+
+    expect(() => cqn4sql(q, model)).to.throw(
+      /The expanded column "author.books.title" must be part of the group by clause/,
+    )
+  })
+  it('deeply nested path not part of group by', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      ID,
+      Books.author { books { author { name } } , ID }
+    } group by author.ID`
+
+    expect(() => cqn4sql(q, model)).to.throw(
+      /The expanded column "author.books.author.name" must be part of the group by clause/,
+    )
+  })
+
+  it('expand path with filter must be an exact match in group by', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      Books.ID,
       author[name='King'] { name }
     } group by author.name`
+
+    expect(() => cqn4sql(q, model)).to.throw(`The expanded column "author[{"ref":["name"]},"=",{"val":"King"}].name" must be part of the group by clause`)
+  })
+  it('expand path with filter must be an exact match in group by (2)', () => {
+    const q = CQL`SELECT from bookshop.Books {
+      Books.ID,
+      author { name }
+    } group by author[name='King'].name`
+
+    expect(() => cqn4sql(q, model)).to.throw(`The expanded column "author.name" must be part of the group by clause`)
   })
 })
 // the tests in here are a copy of the tests in `./inline.test.js`

--- a/hana/lib/HANAService.js
+++ b/hana/lib/HANAService.js
@@ -487,12 +487,33 @@ class HANAService extends SQLService {
                   }
                 })
 
-                x.SELECT.from = {
-                  join: 'inner',
-                  args: [x.SELECT.from, { ref: [parent.alias], as: parent.as }],
-                  on: x.SELECT.where,
-                  as: x.SELECT.from.as,
+                if (x.SELECT.from) {
+                  x.SELECT.from = {
+                    join: 'inner',
+                    args: [x.SELECT.from, { ref: [parent.alias], as: parent.as }],
+                    on: x.SELECT.where,
+                    as: x.SELECT.from.as,
+                  }
+                } else {
+                  x.SELECT.from = { ref: [parent.alias], as: parent.as }
+                  x.SELECT.columns.forEach(col => {
+                    // if (col.ref?.length === 1) { col.ref.unshift(parent.as) }
+                    if (col.ref?.length > 1) {
+                      const colName = this.column_name(col)
+                      if (col.ref[0] !== parent.as) {
+                        // Inject foreign columns into parent select
+                        const as = `$$${col.ref.join('.')}$$`
+                        parent.SELECT.columns.push({ __proto__: col, ref: col.ref, as })
+                        col.as = colName
+                        col.ref = [parent.as, as]
+                      } else if (!parent.SELECT.columns.some(c => this.column_name(c) === colName)) {
+                        // Inject local columns into parent select
+                        parent.SELECT.columns.push({ __proto__: col })
+                      }
+                    }
+                  })
                 }
+
                 x.SELECT.where = undefined
                 x.SELECT.expand = 'root'
                 x.SELECT.parent = parent
@@ -531,6 +552,8 @@ class HANAService extends SQLService {
             },
         )
         .filter(a => a)
+
+      if (sql.length === 0) sql = '*'
 
       if (SELECT.expand === 'root') {
         this._blobs = blobs
@@ -584,6 +607,10 @@ class HANAService extends SQLService {
 
     SELECT_expand(_, sql) {
       return sql
+    }
+
+    from_dummy() {
+      return ' FROM DUMMY'
     }
 
     extractForeignKeys(xpr, alias, foreignKeys = []) {

--- a/test/scenarios/bookshop/read.test.js
+++ b/test/scenarios/bookshop/read.test.js
@@ -39,6 +39,11 @@ describe('Bookshop - Read', () => {
     expect(res.data['@odata.count']).to.be.eq(5)
   })
 
+  test('Books with groupby with path expression and expand result', async () => {
+    const res = await GET('/admin/Books?$apply=filter(title%20ne%20%27bar%27)/groupby((author/name),aggregate(price with sum as totalAmount))', admin)
+    expect(res.data.value.length).to.be.eq(4) // As there are two books which have the same author
+  })
+
   test('Path expression', async () => {
     const q = CQL`SELECT title, author.name as author FROM sap.capire.bookshop.Books where author.name LIKE '%a%'`
     const res = await cds.run(q)

